crypto/src/key/rschnorr/mod.rs

```diff
@@ -155,6 +162,105 @@ impl Derivable for MLRistrettoPrivateKey {
     }
 }
 
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct MLPrivateKey {
+    key_data: secp256k1::SecretKey,
+}
+
+impl zeroize::Zeroize for MLPrivateKey {
+    fn zeroize(&mut self) {
+        // TODO find a safe way to retrieve a mutable slice of the underlying SecretKey data array
+        let key_data = unsafe {
```

Comment:
So, like I mentioned the other day, doing this manually with unsafe is basically playing with fire and quite frankly unnecessary (at this point). Off the top of my head, you can't guarantee the allocation length, which may or may not be a problem.

The issue here is that the slightest mistake will have a tremendous cost in the form memory corruption.

What I suggest doing is disabling this implementation (not necessarily completely removing it), and opening a github issue stating that this has to be studied, justified then done with address sanitizer checks combined with C code review, so that we guarantee we don't do it wrong.

Reply:
I did some digging around how this could be done and here is an idea. Not sure it is viable but may be worth trying out.

The main insight is that the `Zeroize` trait is automatically implemented tor any type that implements `DefaultIsZeroes`. That could be used as follows:

1. Create an internal wrapper over `secp256k1::SecretKey`, with the following impls:
  a) `Copy`, `Clone`, `From<secp256k1::SecretKey>`
  b) `Default` that creates an all-zero private key. Could be done using `secp256k1::SecretKey::from_slice` but there may be a better way
  c) `DefaultIsZeroes`
  d) `Deref(Mut)<Target = secp256k1::SecretKey>` if the amount of unwrapping becomes tedious
2. If the wrapper type is used as `MLPrivateKey::key_data`, it should be enough to just `#[derive(Zeroize)]`

I know the likes of `Default` and `Deref` are somewhat controversial, that's why I suggest the wrapper type to be private or even separated into its own module. Care should be taken it is not exposed outside of this module.

Reply:
Your solution is a good one, Lukas, in the sense that no more unsafe is needed. Though I think we'd need another struct layer on top of that to prevent `Default::default()` from being used by developers... because a default for a private key being 0 is basically playing with fire... and that's an understatement.

Reply:
> Though I think we'd need another struct layer on top of that to prevent `Default::default()` from being used by developers...

That's exactly what I meant when I suggested the wrapper struct should be private/internal and used within `MLPrivateKey` so the users don't have direct access to it.

Reply:
`secp256k1::SecretKey::from_slice` returns an error when the key is zero.

Furthermore if we managed the create a zeroed key, it would be still problematic because when signing with that key a panic would happen.

This is actually why the rust-bitcoin developers decided not to implement zeroize on the secret key.

What we could do, is to implement zeroing on drop. This can be studied on another issue.

Reply:
> `secp256k1::SecretKey::from_slice` returns an error when the key is zero.

Ok, this is an issue which will make my approach with `Default` not viable sadly.